prov/rxm/src/rxm_cq.c

```diff
@@ -491,20 +505,24 @@ ssize_t rxm_cq_handle_large_data(struct rxm_rx_buf *rx_buf)
 				rx_buf, rx_buf->recv_entry->total_len);
 		}
 		total_recv_len -= copy_len;
+
+		RXM_DEC_TX_CREDITS(rx_buf->ep, ret);
+		if (OFI_UNLIKELY(ret)) {
+			ret = rxm_rndv_defer_readv(i, iov, desc, count, rx_buf);
+			if (OFI_UNLIKELY(ret))
+				goto readv_err;
+		}
+
 		ret = fi_readv(rx_buf->conn->msg_ep, iov, desc, count, 0,
 			       rx_buf->rndv_hdr->iov[i].addr,
 			       rx_buf->rndv_hdr->iov[i].key, rx_buf);
 		if (OFI_UNLIKELY(ret)) {
+			RXM_INC_TX_CREDITS(rx_buf->ep);
 			if (OFI_LIKELY(ret == -FI_EAGAIN)) {
-				struct rxm_deferred_tx_entry *def_tx_entry;
-
-				ret = rxm_cq_rndv_read_prepare_deferred(
-						&def_tx_entry, i, iov, desc,
-						count, rx_buf);
-				if (ret)
-					goto readv_err;
-				rxm_ep_enqueue_deferred_tx_queue(def_tx_entry);
-				continue;
+				ret = rxm_rndv_defer_readv(i, iov, desc,
```

Comment:
I'm concerned that having both a credit scheme paired with a deferred queue without locking around updates to both will result in races where items are added to the queue after credits have been updated (or vice-versa), leaving them unprocessed or handled out of order.